File: lib/id3/picture.ex
defmodule ID3.Picture do
  @moduledoc """
  A struct representing a picture in the ID3 tag.
  See https://docs.rs/id3/0.3.0/id3/frame/struct.Picture.html

  ### PictureType
  Due to limitations of `rust-id3`, multiple pictures with same `picture_type` is not available.
  When writing, a picture with the same `picture_type` within the existing tab, will overwrite that picture.
  When reading, `rust-id3` only returns one of the picture of the same type.
  """

  defstruct mime_type: nil, picture_type: nil, description: "", data: nil

  @typedoc """
  A structure representing an ID3 picture frame's contents.

  - `data` is a binary of unsigned char(8bit)s.
  - `picture_type` will be `:Other` otherwise correctly given.
  """
  @type t :: %__MODULE__{
          mime_type: String.t(),
          picture_type: picture_type,
          description: String.t(),
          data: :binary
        }

  @picture_types ~W(
    Other Icon OtherIcon
    CoverFront CoverBack
    Leaflet Media LeadArtist
    Conductor Band Composer Lyricist
    RecordingLocation DuringRecording DuringPerformance
    ScreenCapture BrightFish
    Illustration BandLogo PublisherLogo
  )a

  @typedoc """
  Types of pictures used in APIC frames.
  """
  @type picture_type ::
          :Other
          | :Icon
          | :OtherIcon
          | :CoverFront
          | :CoverBack
          | :Leaflet
          | :Media
          | :LeadArtist
          | :Conductor
          | :Band
          | :Composer
          | :Lyricist
          | :RecordingLocation
          | :DuringRecording
          | :DuringPerformance
          | :ScreenCapture
          | :BrightFish
          | :Illustration
          | :BandLogo
          | :PublisherLogo

  @doc """
  Creates a new Picture.

  Needs the data binary and the MIME type of the data.

  ## Options
  - `picture_type`: ID3 tags can have a picture for each type. One of `t:Picture.picture_type/0`.
  - `description`: Text description about the picture.

  ## Examples
      iex> data = File.read!("full/path/to/foo.jpg")
      iex> ID3.Picture.new(data, "image/jpeg", picture_type: :CoverFront)
      {:ok,
        %ID3.Picture{
          data: <<255, 167, ...>>,
          description: "",
          mime_type: "image/jpeg",
          picture_type: :CoverFront
        }
      }
  """
  @spec new(binary, String.t(), options :: [option]) :: {:ok, Picture.t()} | :error
        when option: {:picture_type, picture_type()} | {:description, String.t()}
  def new(data, mime_type, options \\ [])
      when data |> is_binary and mime_type |> is_binary and options |> is_list do
    picture_type = options |> Keyword.get(:picture_type, :Other)
    description = options |> Keyword.get(:description, "")

    with true <- @picture_types |> Enum.member?(picture_type),
         true <- description |> String.valid?() do
      picture = %__MODULE__{
        picture_type: picture_type,
        mime_type: mime_type,
        description: "",
        data: data
      }

      {:ok, picture}
    else
      _ -> :error
    end
  end
end

File: native/id3/src/nif_converter/picture.rs
use crate::NifBinary;
use id3::frame::{Picture, PictureType};
use rustler::types::atom::Atom as NifAtom;
use rustler::NifStruct;

mod atoms {
    use rustler::rustler_atoms;

    rustler_atoms! {
        atom ok;
        atom error;
    }
}

#[allow(non_snake_case)]
mod pictype_atoms {
    use rustler::rustler_atoms;

    rustler_atoms! {
        atom Other;
        atom Icon;
        atom OtherIcon;
        atom CoverFront;
        atom CoverBack;
        atom Leaflet;
        atom Media;
        atom LeadArtist;
        atom Artist;
        atom Conductor;
        atom Band;
        atom Composer;
        atom Lyricist;
        atom RecordingLocation;
        atom DuringRecording;
        atom DuringPerformance;
        atom ScreenCapture;
        atom BrightFish;
        atom Illustration;
        atom BandLogo;
        atom PublisherLogo;
    }
}

#[derive(NifStruct)]
#[module = "ID3.Picture"]
/// Struct matching Elixir ID3's `Picture` struct.
pub struct ID3Picture {
    pub mime_type: String,
    pub picture_type: NifAtom,
    pub description: String,
    pub data: NifBinary, // not a valid String, but a elixir binary.
}

impl From<&Picture> for ID3Picture {
    fn from(pic: &Picture) -> Self {
        Self {
            mime_type: pic.mime_type.clone(),
            picture_type: from_picture_type(pic.picture_type),
            description: pic.description.clone(),
            data: NifBinary(pic.data.clone()),
        }
    }
}

impl From<ID3Picture> for Picture {
    fn from(pic: ID3Picture) -> Self {
        Self {
            mime_type: pic.mime_type,
            picture_type: to_picture_type(pic.picture_type),
            description: pic.description,
            data: pic.data.vec(), //unsafe { pic.data.clone().as_mut_vec().to_vec() },
        }
    }
}

fn from_picture_type(pic: PictureType) -> NifAtom {
    use id3::frame::PictureType::*;

    match pic {
        Icon => pictype_atoms::Icon(),
        OtherIcon => pictype_atoms::OtherIcon(),
        CoverFront => pictype_atoms::CoverFront(),
        CoverBack => pictype_atoms::CoverBack(),
        Leaflet => pictype_atoms::Leaflet(),
        Media => pictype_atoms::Media(),
        LeadArtist => pictype_atoms::LeadArtist(),
        Artist => pictype_atoms::Artist(),
        Conductor => pictype_atoms::Conductor(),
        Band => pictype_atoms::Band(),
        Composer => pictype_atoms::Composer(),
        Lyricist => pictype_atoms::Lyricist(),
        RecordingLocation => pictype_atoms::RecordingLocation(),
        DuringRecording => pictype_atoms::DuringRecording(),
        DuringPerformance => pictype_atoms::DuringPerformance(),
        ScreenCapture => pictype_atoms::ScreenCapture(),
        BrightFish => pictype_atoms::BrightFish(),
        Illustration => pictype_atoms::Illustration(),
        BandLogo => pictype_atoms::BandLogo(),
        PublisherLogo => pictype_atoms::PublisherLogo(),
        _ => pictype_atoms::Other(),
    }
}

fn to_picture_type(atom: NifAtom) -> PictureType {
    if atom == pictype_atoms::Icon() {
        PictureType::Icon
    } else if atom == pictype_atoms::OtherIcon() {
        PictureType::OtherIcon
    } else if atom == pictype_atoms::CoverFront() {
        PictureType::CoverFront
    } else if atom == pictype_atoms::CoverBack() {
        PictureType::CoverBack
    } else if atom == pictype_atoms::Leaflet() {
        PictureType::Leaflet
    } else if atom == pictype_atoms::Media() {
        PictureType::Media
    } else if atom == pictype_atoms::LeadArtist() {
        PictureType::LeadArtist
    } else if atom == pictype_atoms::Artist() {
        PictureType::Artist
    } else if atom == pictype_atoms::Conductor() {
        PictureType::Conductor
    } else if atom == pictype_atoms::Band() {
        PictureType::Band
    } else if atom == pictype_atoms::Composer() {
        PictureType::Composer
    } else if atom == pictype_atoms::Lyricist() {
        PictureType::Lyricist
    } else if atom == pictype_atoms::RecordingLocation() {
        PictureType::RecordingLocation
    } else if atom == pictype_atoms::DuringRecording() {
        PictureType::DuringRecording
    } else if atom == pictype_atoms::DuringPerformance() {
        PictureType::DuringPerformance
    } else if atom == pictype_atoms::ScreenCapture() {
        PictureType::ScreenCapture
    } else if atom == pictype_atoms::BrightFish() {
        PictureType::BrightFish
    } else if atom == pictype_atoms::Illustration() {
        PictureType::Illustration
    } else if atom == pictype_atoms::BandLogo() {
        PictureType::BandLogo
    } else if atom == pictype_atoms::PublisherLogo() {
        PictureType::PublisherLogo
    } else {
        PictureType::Other
    }
}
